File: tools/run_state/validate.py
"""Schema validation for .prove/runs artifact JSON files.

Thin wrapper around the DSL-based validator from ``tools/schema/validate.py``
parameterized by the per-kind schemas in ``schemas.py``.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from tools.run_state.schemas import SCHEMA_BY_KIND, infer_kind
from tools.schema.validate import ValidationError, validate_config


def validate_data(
    data: Any, kind: str, strict: bool = False
) -> list[ValidationError]:
    """Validate a parsed JSON value against the schema for ``kind``."""
    if kind not in SCHEMA_BY_KIND:
        return [ValidationError("", f"unknown schema kind: {kind!r}")]
    if not isinstance(data, dict):
        return [ValidationError("", "top-level value must be a JSON object")]
    return validate_config(data, SCHEMA_BY_KIND[kind], strict=strict)


def validate_file(
    path: str | Path, kind: str | None = None, strict: bool = False
) -> tuple[dict | None, list[ValidationError]]:
    """Validate a run-state JSON file.

    If ``kind`` is None, infer from the filename (``prd.json``, ``plan.json``,
    ``state.json``, or ``reports/*.json``).
    """
    p = Path(path)
    if not p.exists():
        return None, [ValidationError(str(p), "file not found")]

    try:
        with open(p, encoding="utf-8") as f:
            data = json.load(f)
    except json.JSONDecodeError as e:
        return None, [ValidationError(str(p), f"invalid JSON: {e}")]

    if kind is None:
        kind = infer_kind(str(p))
        if kind is None:
            return data, [
                ValidationError(
                    str(p),
                    "cannot infer schema kind from filename — pass --kind explicitly",
                )
            ]

    errors = validate_data(data, kind, strict=strict)
    return data, errors

File: packages/cli/src/topics/run-state/validate.ts
/**
 * Schema validation for `.prove/runs/` artifact JSON files.
 *
 * Thin wrapper around `validator-engine.ts`, parameterized by the per-kind
 * schemas in `schemas.ts`. Ported from `tools/run_state/validate.py`.
 *
 * Return-shape divergence (deliberate): the Python source returns
 * `list[ValidationError]` / `tuple[dict, list[ValidationError]]`; the TS
 * port returns the envelope `{ ok, kind, version, errors }` so downstream
 * hooks can branch on `ok` without introspecting an error object. The error
 * *strings* inside `errors` match Python byte-for-byte — agents see the same
 * stderr wording.
 */

import { readFileSync } from 'node:fs';
import { CURRENT_SCHEMA_VERSION, SCHEMA_BY_KIND, inferKind } from './schemas';
import { type Schema, ValidationError, validateConfig } from './validator-engine';

export interface ValidateResult {
  ok: boolean;
  kind: string;
  version: string;
  /** Stringified findings in `"  ERROR: <path>: <message>"` format. */
  errors: string[];
}

/**
 * Validate a parsed JSON value against the schema for `kind`.
 *
 * `strict: true` promotes unknown-key warnings into errors (matches
 * Python behavior in `_validator.validate_config`).
 */
export function validateData(data: unknown, kind: string, strict = false): ValidateResult {
  if (!(kind in SCHEMA_BY_KIND)) {
    return {
      ok: false,
      kind,
      version: CURRENT_SCHEMA_VERSION,
      errors: [new ValidationError('', `unknown schema kind: '${kind}'`).toString()],
    };
  }
  if (!isPlainObject(data)) {
    return {
      ok: false,
      kind,
      version: CURRENT_SCHEMA_VERSION,
      errors: [new ValidationError('', 'top-level value must be a JSON object').toString()],
    };
  }
  const schema = SCHEMA_BY_KIND[kind] as Schema;
  const findings = validateConfig(data, schema, strict);
  const hardErrors = findings.filter((e) => e.severity === 'error');
  return {
    ok: hardErrors.length === 0,
    kind,
    version: schema.version,
    errors: findings.map((e) => e.toString()),
  };
}

/**
 * Read a JSON file from disk and validate it. When `kind` is omitted, the
 * schema is inferred from the filename (`prd.json`, `plan.json`,
 * `state.json`, `reports/*.json`).
 *
 * Returns the same `{ ok, kind, version, errors }` shape as `validateData`.
 * On file-not-found or invalid JSON, `kind` is echoed as the inferred value
 * (or an empty string if nothing could be inferred) and `errors` carries the
 * I/O diagnostic.
 */
export function validateFile(path: string, kind?: string, strict = false): ValidateResult {
  let raw: string;
  try {
    raw = readFileSync(path, 'utf8');
  } catch {
    return {
      ok: false,
      kind: kind ?? '',
      version: CURRENT_SCHEMA_VERSION,
      errors: [new ValidationError(path, 'file not found').toString()],
    };
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    return {
      ok: false,
      kind: kind ?? '',
      version: CURRENT_SCHEMA_VERSION,
      errors: [new ValidationError(path, `invalid JSON: ${msg}`).toString()],
    };
  }

  const resolvedKind = kind ?? inferKind(path);
  if (!resolvedKind) {
    return {
      ok: false,
      kind: '',
      version: CURRENT_SCHEMA_VERSION,
      errors: [
        new ValidationError(
          path,
          'cannot infer schema kind from filename — pass --kind explicitly',
        ).toString(),
      ],
    };
  }

  return validateData(parsed, resolvedKind, strict);
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}
